In [29]:
import requests
import json

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly
from plotly.offline import download_plotlyjs, init_notebook_mode,plot,iplot

In [28]:
pip install seaborn

                                              0.0/294.8 kB ? eta -:--:--
     ----                                    30.7/294.8 kB 1.3 MB/s eta 0:00:01
     --------                              71.7/294.8 kB 787.7 kB/s eta 0:00:01
     ------------------                     143.4/294.8 kB 1.1 MB/s eta 0:00:01
     ------------------------------         235.5/294.8 kB 1.4 MB/s eta 0:00:01
     -------------------------------------- 294.8/294.8 kB 1.5 MB/s eta 0:00:00
Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 23.1.2 -> 23.3.2
[notice] To update, run: python.exe -m pip install --upgrade pip


In [2]:
df = pd.read_csv(r'C:\Users\Dorothy\OneDrive\Documents\Nairobi County Healthcare Accessibility\Nairobi-County-Accessibility-of-Health-Care\data\ehealth-kenya-facilities-download-21102015.csv')
nairobi_dataset = df[df['County'] == "Nairobi"]
nairobi_dataset.head()

Facility Code                                Facility Name Province  \
1           19310   St Jude's Huruma Community Health Services  Nairobi   
7           13043                        7Kr Mrs Health Centre  Nairobi   
10          20346                      AAR Adams Health Centre  Nairobi   
11          12861                       AAR City Centre Clinic  Nairobi   
12          16796          AAR Clinic Sarit Centre (Westlands)  Nairobi   

     County   District   Division            Type  \
1   Nairobi    Mathare     Huruma  Medical Clinic   
7   Nairobi    Langata   Lang'ata   Health Centre   
10  Nairobi      Kibra     Woodly  Medical Clinic   
11  Nairobi    Starehe    Starehe  Medical Clinic   
12  Nairobi  Westlands  Parklands  Medical Clinic   

                                      Owner                   Location  \
1            Private Practice - Unspecified                     Huruma   
7                              Armed Forces                  Mugumoini   
10  Private Practice - General Practitioner                     Woodly   
11         Private Enterprise (Institution)  Central Business District   
12         Private Enterprise (Institution)                  Parklands   

       Sub Location  ...  IPD OPD OUTREACH  PMTCT  RAD/XRAY RHTC/RHDC TB DIAG  \
1            Huruma  ...  NaN NaN      NaN    NaN       NaN       NaN     NaN   
7         Mugumoini  ...    Y NaN      NaN    NaN       NaN       NaN     NaN   
10           Woodly  ...  NaN NaN      NaN    NaN       NaN       NaN     NaN   
11              NaN  ...    Y NaN      NaN    NaN       NaN       NaN     NaN   
12  Upper Parklands  ...  NaN NaN      NaN    NaN       NaN       NaN     NaN   

   TB LABS TB TREAT YOUTH  
1      NaN      NaN   NaN  
7      NaN      NaN   NaN  
10     NaN      NaN   NaN  
11     NaN      NaN   NaN  
12     NaN      NaN   NaN  

[5 rows x 50 columns]

In [3]:
# column names
df.columns.tolist()

['Facility Code',
 'Facility Name',
 'Province',
 'County',
 'District',
 'Division',
 'Type',
 'Owner',
 'Location',
 'Sub Location',
 'Description of Location',
 'Constituency',
 'Nearest Town',
 'Beds',
 'Cots',
 'Official Landline',
 'Official Fax',
 'Official Mobile',
 'Official Email',
 'Official Address',
 'Official Alternate No',
 'Town',
 'Post Code',
 'In Charge',
 'Job Title of in Charge',
 'Open 24 Hours',
 'Open Weekends',
 'Operational Status',
 'ANC',
 'ART',
 'BEOC',
 'BLOOD',
 'CAES SEC',
 'CEOC',
 'C-IMCI',
 'EPI',
 'FP',
 'GROWM',
 'HBC',
 'HCT',
 'IPD',
 'OPD',
 'OUTREACH',
 'PMTCT',
 'RAD/XRAY',
 'RHTC/RHDC',
 'TB DIAG',
 'TB LABS',
 'TB TREAT',
 'YOUTH']

In [4]:
# number of data points
print(f" Number of rows:  {df.shape[0]} \n Number of columns: {df.shape[1]} \n")

print("Narobi County Dataset: " )
print(f" Number of rows:  {nairobi_dataset.shape[0]} \n Number of columns: {nairobi_dataset.shape[1]} ")

 Number of rows:  10505 
 Number of columns: 50 

Narobi County Dataset: 
 Number of rows:  942 
 Number of columns: 50 


In [5]:
df.describe()

Facility Code          Beds          Cots  Official Alternate No  ANC  \
count   10505.000000  10418.000000  10428.000000           9.520000e+02  0.0   
mean    15726.711471      5.432905      0.744342           6.695277e+08  NaN   
std      3246.324955     29.862656     11.332022           5.568178e+08  NaN   
min     10001.000000      0.000000      0.000000           0.000000e+00  NaN   
25%     12957.000000      0.000000      0.000000           7.138962e+08  NaN   
50%     15702.000000      0.000000      0.000000           7.227692e+08  NaN   
75%     18569.000000      1.000000      0.000000           7.278981e+08  NaN   
max     21281.000000   1455.000000   1000.000000           7.382217e+09  NaN   

       BEOC  BLOOD  CAES SEC  CEOC  EPI  ...  HCT  OPD  OUTREACH  PMTCT  \
count   0.0    0.0       0.0   0.0  0.0  ...  0.0  0.0       0.0    0.0   
mean    NaN    NaN       NaN   NaN  NaN  ...  NaN  NaN       NaN    NaN   
std     NaN    NaN       NaN   NaN  NaN  ...  NaN  NaN       NaN    NaN   
min     NaN    NaN       NaN   NaN  NaN  ...  NaN  NaN       NaN    NaN   
25%     NaN    NaN       NaN   NaN  NaN  ...  NaN  NaN       NaN    NaN   
50%     NaN    NaN       NaN   NaN  NaN  ...  NaN  NaN       NaN    NaN   
75%     NaN    NaN       NaN   NaN  NaN  ...  NaN  NaN       NaN    NaN   
max     NaN    NaN       NaN   NaN  NaN  ...  NaN  NaN       NaN    NaN   

       RAD/XRAY  RHTC/RHDC  TB DIAG  TB LABS  TB TREAT  YOUTH  
count       0.0        0.0      0.0      0.0       0.0    0.0  
mean        NaN        NaN      NaN      NaN       NaN    NaN  
std         NaN        NaN      NaN      NaN       NaN    NaN  
min         NaN        NaN      NaN      NaN       NaN    NaN  
25%         NaN        NaN      NaN      NaN       NaN    NaN  
50%         NaN        NaN      NaN      NaN       NaN    NaN  
75%         NaN        NaN      NaN      NaN       NaN    NaN  
max         NaN        NaN      NaN      NaN       NaN    NaN  

[8 rows x 21 columns]

In [6]:
nairobi_dataset.describe()

Facility Code         Beds        Cots  Official Alternate No  ANC  \
count     942.000000   942.000000  942.000000           1.570000e+02  0.0   
mean    16799.744161     7.787686    1.152866           6.498035e+08  NaN   
std      3129.912948    58.966655   15.196691           5.890122e+08  NaN   
min     12861.000000     0.000000    0.000000           2.222575e+06  NaN   
25%     13112.250000     0.000000    0.000000           7.103701e+08  NaN   
50%     18346.000000     0.000000    0.000000           7.225878e+08  NaN   
75%     19492.750000     1.000000    0.000000           7.288983e+08  NaN   
max     21281.000000  1455.000000  427.000000           7.329109e+09  NaN   

       BEOC  BLOOD  CAES SEC  CEOC  EPI  ...  HCT  OPD  OUTREACH  PMTCT  \
count   0.0    0.0       0.0   0.0  0.0  ...  0.0  0.0       0.0    0.0   
mean    NaN    NaN       NaN   NaN  NaN  ...  NaN  NaN       NaN    NaN   
std     NaN    NaN       NaN   NaN  NaN  ...  NaN  NaN       NaN    NaN   
min     NaN    NaN       NaN   NaN  NaN  ...  NaN  NaN       NaN    NaN   
25%     NaN    NaN       NaN   NaN  NaN  ...  NaN  NaN       NaN    NaN   
50%     NaN    NaN       NaN   NaN  NaN  ...  NaN  NaN       NaN    NaN   
75%     NaN    NaN       NaN   NaN  NaN  ...  NaN  NaN       NaN    NaN   
max     NaN    NaN       NaN   NaN  NaN  ...  NaN  NaN       NaN    NaN   

       RAD/XRAY  RHTC/RHDC  TB DIAG  TB LABS  TB TREAT  YOUTH  
count       0.0        0.0      0.0      0.0       0.0    0.0  
mean        NaN        NaN      NaN      NaN       NaN    NaN  
std         NaN        NaN      NaN      NaN       NaN    NaN  
min         NaN        NaN      NaN      NaN       NaN    NaN  
25%         NaN        NaN      NaN      NaN       NaN    NaN  
50%         NaN        NaN      NaN      NaN       NaN    NaN  
75%         NaN        NaN      NaN      NaN       NaN    NaN  
max         NaN        NaN      NaN      NaN       NaN    NaN  

[8 rows x 21 columns]

In [7]:

def get_value_at_max(dataframe, column_max, column_extract):
    max_index = dataframe[column_max].idxmax()
    value_at_max = dataframe.at[max_index, column_extract]
    return value_at_max

nation_largest_hospital = get_value_at_max(df, 'Beds', 'Facility Name')
nation_largest_maternity_hospital = get_value_at_max(df, 'Cots', 'Facility Name')
nairoby_largest_maternity_hospital = get_value_at_max(nairobi_dataset, 'Cots', 'Facility Name')
print( f"The country's largest hospital by number of beds is :{nation_largest_hospital} \nHospital with the most number of cots in the country  is :{nation_largest_maternity_hospital} \nHospital with the most number of cots in Nairobi County is : {nairoby_largest_maternity_hospital}" )
#    print("The country's largest maternity hospital by number of beds is :", nation_largest_maternity_hospital) 
# "Nairobu's largest maternity hospital by number of beds is :", nation_largest_maternity_hospital)

The country's largest hospital by number of beds is :Kenyatta National Hospital 
Hospital with the most number of cots in the country  is :Njerian Dispensary 
Hospital with the most number of cots in Nairobi County is : Kenyatta National Hospital


In [8]:
#number of unique values per column
unique_values=pd.DataFrame(nairobi_dataset.apply(lambda x: len(x.value_counts(dropna=False)), axis=0), 
                           columns=['Unique Value Count']).sort_values(by='Unique Value Count', ascending=True)
unique_values

Unique Value Count
YOUTH                                     1
TB TREAT                                  1
ANC                                       1
BEOC                                      1
BLOOD                                     1
CAES SEC                                  1
CEOC                                      1
EPI                                       1
HCT                                       1
GROWM                                     1
RHTC/RHDC                                 1
OUTREACH                                  1
TB LABS                                   1
Province                                  1
PMTCT                                     1
County                                    1
RAD/XRAY                                  1
OPD                                       1
TB DIAG                                   1
ART                                       2
IPD                                       2
FP                                        2
HBC                                       2
C-IMCI                                    2
Open 24 Hours                             3
Operational Status                        3
Open Weekends                             3
Job Title of in Charge                    6
Cots                                     15
Constituency                             17
District                                 17
Official Fax                             20
Type                                     20
Owner                                    21
Town                                     29
Beds                                     50
Post Code                                51
Division                                 90
Location                                154
Official Alternate No                   156
Official Landline                       194
Sub Location                            216
Official Email                          219
Official Mobile                         299
Official Address                        330
Nearest Town                            364
In Charge                               575
Description of Location                 625
Facility Name                           941
Facility Code                           942

In [9]:
# the percentage of missing values in the nairobi dataset
def missing_values(df):    
    totalCells = df.size   
    missingCount = df.isnull().sum()
    totalMissing = missingCount.sum()    
    print("There are a total of", totalCells, "datapoints")
    print("The dataset contains", totalMissing, "missing values.")
    print(round(((totalMissing/totalCells) * 100), 2), "%", "of the datapoints are missing")

missing_values(nairobi_dataset)

There are a total of 47100 datapoints
The dataset contains 26633 missing values.
56.55 % of the datapoints are missing


In [10]:
def columns_with_most_missing_values(dataframe, num_columns): 
     # Calculate the percentage of missing values in each column
    missing_percentage = (dataframe.isnull().sum() / len(dataframe)) * 100    
    # Sort columns by their missing percentage in descending order
    sorted_columns = missing_percentage.sort_values(ascending=False)    
    # Get the top 'num_columns' columns with the most missing values
    columns_most_missing = sorted_columns.head(num_columns)    
    return columns_most_missing

columns_most_missing = columns_with_most_missing_values(nairobi_dataset, 25)
print("Columns with the most missing values and percentage of missing values:\n",columns_most_missing)

Columns with the most missing values and percentage of missing values:
 YOUTH                    100.000000
CEOC                     100.000000
GROWM                    100.000000
BEOC                     100.000000
OPD                      100.000000
OUTREACH                 100.000000
BLOOD                    100.000000
HCT                      100.000000
CAES SEC                 100.000000
PMTCT                    100.000000
RAD/XRAY                 100.000000
RHTC/RHDC                100.000000
EPI                      100.000000
TB DIAG                  100.000000
TB LABS                  100.000000
TB TREAT                 100.000000
ANC                      100.000000
Official Fax              97.983015
C-IMCI                    92.356688
ART                       88.428875
Official Alternate No     83.333333
Official Landline         79.087049
Official Email            76.433121
HBC                       73.460722
FP                        70.276008
dtype: float64


Majority of the above columns are fields that highlight the type of tests offered by each hospital. Since most of them are empty, these columns will be removed from the dataset. 

In [15]:
def remove_columns_with_high_missing_values(dataframe, threshold=70):
    missing_percentage = (dataframe.isnull().sum() / len(dataframe)) * 100        
    columns_to_remove = missing_percentage[missing_percentage > threshold].index      
    dataframe_filtered = dataframe.drop(columns=columns_to_remove)    
    return dataframe_filtered

nairobi_dataset_filtered = remove_columns_with_high_missing_values(nairobi_dataset_filtered, threshold=70)
print("DataFrame after removing columns with over 70% missing values: ",nairobi_dataset_filtered.shape)


DataFrame after removing columns with over 70% missing values:  (942, 25)


In [17]:
missing_values(nairobi_dataset_filtered)

There are a total of 23550 datapoints
The dataset contains 4389 missing values.
18.64 % of the datapoints are missing


By removing columns with removing columns with over 70% missing values, the number of columns have reduced by 50% and the missing datapoints reduced from 56.55% to 18.64%. 

#### Observations from Initial Analysis

1. There are a total of 10505 facilities countrywide and 942 facilities in Nairobi County.
2. On average there are approximately 7 beds per hospital within Nairobi which is higher than the national average of 5 beds per hospital.
3. Kenyatta National Hospital has the most number of beds(1455) and cots(427) within Nairobi County, while Njerian Dispensary in Bomet County has the most number of cots (1000)country wide.



### Exploratory Data Analysis

An analysis on the distribution of hospitals in Nairobi County by geographical location, Ownership, Quality of services, Skills of personnel and Operationality


In [ ]:
def plot_hist(df:pd.DataFrame, column:str, color:str)->None:
    # plt.figure(figsize=(15, 10))
    # fig, ax = plt.subplots(1, figsize=(12, 7))
    sns.displot(data=df, x=column, color=color, kde=True, height=7, aspect=2)
    plt.title(f'Distribution of {column}', size=20, fontweight='bold')
    plt.show()

def plot_count(df:pd.DataFrame, column:str) -> None:
    plt.figure(figsize=(12, 7))
    sns.countplot(data=df, x=column)
    plt.title(f'Distribution of {column}', size=20, fontweight='bold')
    plt.show()
    
def plot_bar(df:pd.DataFrame, x_col:str, y_col:str, title:str, xlabel:str, ylabel:str)->None:
    plt.figure(figsize=(12, 7))
    sns.barplot(data = df, x=x_col, y=y_col)
    plt.title(title, size=20)
    plt.xticks(rotation=75, fontsize=14)
    plt.yticks( fontsize=14)
    plt.xlabel(xlabel, fontsize=16)
    plt.ylabel(ylabel, fontsize=16)
    plt.show()

def plot_heatmap(df:pd.DataFrame, title:str, cbar=False)->None:
    plt.figure(figsize=(12, 7))
    sns.heatmap(df, annot=True, cmap='viridis', vmin=0, vmax=1, fmt='.2f', linewidths=.7, cbar=cbar )
    plt.title(title, size=18, fontweight='bold')
    plt.show()

def plot_box(df:pd.DataFrame, x_col:str, title:str) -> None:
    plt.figure(figsize=(12, 7))
    sns.boxplot(data = df, x=x_col)
    plt.title(title, size=20)
    plt.xticks(rotation=75, fontsize=14)
    plt.show()

def plot_box_multi(df:pd.DataFrame, x_col:str, y_col:str, title:str) -> None:
    plt.figure(figsize=(12, 7))
    sns.boxplot(data = df, x=x_col, y=y_col)
    plt.title(title, size=20)
    plt.xticks(rotation=75, fontsize=14)
    plt.yticks( fontsize=14)
    plt.show()

def plot_scatter(df: pd.DataFrame, x_col: str, y_col: str, title: str, hue: str, style: str) -> None:
    plt.figure(figsize=(12, 7))
    sns.scatterplot(data = df, x=x_col, y=y_col, hue=hue, style=style)
    plt.title(title, size=20)
    plt.xticks(fontsize=14)
    plt.yticks( fontsize=14)
    plt.show()

def correlation(df:pd.DataFrame):
    corr=df.corr()
    matrix=np.triu(df.corr())
    sns.heatmap(corr,annot=True,square=True,mask=matrix)
    fig=plt.gcf()
    figsize=fig.get_size_inches()
    fig.set_size_inches(figsize*4)
    plt.show()

In [13]:

response = requests.get("https://openafrica.net/api/3/action/datastore_search?resource_id=9b94fe50-9d75-4b92-be00-6354c6e6cc88&q=nairobi")
print(response.text)

{"help": "https://openafrica.net/api/3/action/help_show?name=datastore_search", "success": false, "error": {"message": "Not found: Resource \"9b94fe50-9d75-4b92-be00-6354c6e6cc88\" was not found.", "__type": "Not Found Error"}}


In [14]:

# Load JSON data
data = json.loads(response.text)

# Accessing the 'records' key and iterating through the values
for record in data['result']['records']:
    print(record)


KeyError: 'result'